### ISSUE TITLE
Bad error messages when given invalid parameter values from command-line

### ISSUE THREAD

**Issue by akisaarinen** [scallop/scallop#22]
Error messages when parameter value conversions fail are confusing.

For example, given a test application which only accepts numeric 'apples', we get the following output when giving non-numeric 'apples':

```
$ java -jar app.jar --apples asdf
Error: Failed to parse the trailing argument list: 'asdf'
  -a, --apples  <arg>   Number of apples 
```

I would like to see an error message, which:
- tells the user which parameter it was that failed (in this case `--apples`)
- tells the user why the given parameter failed (in this case the value conversion from `asdf` couldn't be done, because it's non-numeric).

The example app I used is this:

``` scala
import org.rogach.scallop.{ScallopConf, Scallop}
import org.rogach.scallop.exceptions.Help

class ExampleConfig(
  arguments: Array[String],
  onError:   (Throwable, Scallop) => Nothing
) extends ScallopConf(arguments) {
  val apple = opt[Int](
    name     = "apples",
    descr    = "Number of apples",
    required = true
  )
  override protected def onError(e: Throwable) = onError(e, builder)
}

object TestApp extends App {
  val config = new ExampleConfig(args, onError)
  println("Number of apples: %d".format(config.apple()))

  private def onError(e: Throwable, scallop: Scallop) = e match {
    case Help =>
      scallop.printHelp
      sys.exit(0)
    case _ =>
      println("Error: %s".format(e.getMessage))
      scallop.printHelp
      sys.exit(1)
  }
}
```

I think the issues lies in the fact that `ValueConverter#parse` returns of type `Either[Unit,Option[A]]`. This means that any information about the error is lost.

Maybe ValueConverters could return something like `Either[Error, Option[A]]` instead, where `Error` would describe the thing that caused conversion to fail? Then value converters could be implemented so that better error messages are given.

For now, there is a workaround: to create value converters which throw an exception instead. So I can do for example the following (other code being the same as above):

``` scala
class ExampleConfig(
  arguments: Array[String],
  onError:   (Throwable, Scallop) => Nothing
) extends ScallopConf(arguments) {
  private implicit val intValueConverter = ScallopConverters.singleArg[Int](_.toInt)
  val apple = opt[Int](
    name     = "apples",
    descr    = "Number of apples",
    required = true
  )
  override protected def onError(e: Throwable) = onError(e, builder)
}

object ScallopConverters {
  def singleArg[A](conv: String => A)(implicit m: Manifest[A]): ValueConverter[A] = new ValueConverter[A] {
    override def parse(s: List[(String, List[String])]) = {
      s match {
        case (param, i :: Nil) :: Nil => {
          Right(Some(conv(i)))
        }
        case Nil => Right(None)
        case _ => Left(Unit)
      }
    }
    val manifest = m
    val argType = ArgType.SINGLE
  }
}
```

This value converter throws an exception and gives a slightly better error message (still not an optimal one):

```
$ java -jar app.jar --apples asdf

Error: For input string: "asdf"
  -a, --apples  <arg>   Number of apples 
```

**Comment by Rogach** [comment]
Well, consider this example:

``` scala
val conf = new ScallopConf(Seq("-a","asdf","-b","2")) {
  val apples = opt[Int]()
  val bananas = opt[Int]()
}
```

It prints, as you wanted, 

```
Wrong format for option 'apples': asdf
```

The problem with your example is that the failed option is last, and parsing engine has to consider trailing arguments _and_ last non-trailing option together - to enable the possibility of list arguments. For example, see this simple case:

```
val conf = new ScallopConf(args) {
  val numbers = opt[List[Int]]()
  val words = trailArg[List[String]]()
}
./prog --numbers 1 2 3 one two three
```

In your case, there are no trailing arguments, so I can try to short-circuit the parsing logic to get you the right message. But in general case, the last argument will always fail with "Failed to parse trailing arguments".

**Comment by Rogach** [comment]
Well, seems I was a bit wrong. I managed to short-circuit both flag and single-arg options in any conditions, so 
now you should be able to get the proper error message. See 761ac32071.

**Comment by akisaarinen** [comment]
Option name guessing seems to be failing in 761ac32?

```
[info] OptionNameGuessing:
[info] - simple
[info] - tricky *** FAILED ***
[info]   org.rogach.scallop.exceptions.IdenticalOptionNames: Option identifier 'aaa' is not unique
[info]   at org.rogach.scallop.Scallop$$anonfun$verify$4.apply(Scallop.scala:479)
[info]   at org.rogach.scallop.Scallop$$anonfun$verify$4.apply(Scallop.scala:479)
[info]   at scala.collection.immutable.Map$Map1.foreach(Map.scala:118)
[info]   at org.rogach.scallop.Scallop.verify(Scallop.scala:478)
[info]   at org.rogach.scallop.ScallopConf.verify(ScallopConf.scala:228)
[info]   at org.rogach.scallop.ScallopConf.afterInit(ScallopConf.scala:358)
[info]   at org.rogach.scallop.AfterInit$class.delayedInit(AfterInit.scala:14)
[info]   at org.rogach.scallop.ScallopConf.delayedInit(ScallopConf.scala:19)
[info]   at OptionNameGuessing$$anonfun$2$Conf$5$.<init>(OptionNameGuessing.scala:18)
[info]   at OptionNameGuessing$$anonfun$2.Conf$2(OptionNameGuessing.scala:18)
```

Does this happen for you also?

**Comment by Rogach** [comment]
Hm. Strange, I run test on my copy and it passes. I'll investigate now.

**Comment by Rogach** [comment]
Well, it does fail on fresh repo clone. That's quite... strange.

**Comment by Rogach** [comment]
Seems there is something directory-dependent - it works under `/home/platon/sync/scallop/` fine (even with fresh copy), but under other directories it fails.

Thanks for pointing this out - I would have never found it otherwise!

**Comment by Rogach** [comment]
Now I can't get that error again, in any directories. Seems that there are some strange issues with sbt.

**Comment by Rogach** [comment]
I wiped ivy cache, re-cloned a fresh project copy, and ran `sbt test` - all tests passed fine. I use sbt 0.12.0.

Does that error still happen for you?

**Comment by akisaarinen** [comment]
I tried cleaning up my IVY cache: after that tests pass once just fine.

But, after doing a few clean, test and restart sbt -cycles, they are failing again, this time by exiting the whole SBT:

`[scallop] Error: Option identifier 'aaa' is not unique`

I guess that would be an unexpected call to `sys.exit` in the tests.

After a few more clean&test -cycles, I get the original error (without exiting SBT):

```
[info] OptionNameGuessing:
[info] - simple
[info] - tricky *** FAILED ***
[info]   org.rogach.scallop.exceptions.IdenticalOptionNames: Option identifier 'aaa' is not unique
[info]   at org.rogach.scallop.Scallop$$anonfun$verify$4.apply(Scallop.scala:479)
[info]   at org.rogach.scallop.Scallop$$anonfun$verify$4.apply(Scallop.scala:479)
[info]   at scala.collection.immutable.Map$Map1.foreach(Map.scala:118)
[info]   at org.rogach.scallop.Scallop.verify(Scallop.scala:478)
[info]   at org.rogach.scallop.ScallopConf.verify(ScallopConf.scala:228)
[info]   at org.rogach.scallop.ScallopConf.afterInit(ScallopConf.scala:358)
[info]   at org.rogach.scallop.AfterInit$class.delayedInit(AfterInit.scala:14)
[info]   at org.rogach.scallop.ScallopConf.delayedInit(ScallopConf.scala:19)
[info]   at OptionNameGuessing$$anonfun$2$Conf$5$.<init>(OptionNameGuessing.scala:18)
[info]   at OptionNameGuessing$$anonfun$2.Conf$2(OptionNameGuessing.scala:18)
```

There is something wrong with this, I don't know exactly what. You could try doing clean / test / restart-sbt -iterations to see if this problem pops up for you.

For the record, I am using SBT 0.12.0, OSX 10.7.5 and Java 1.7.0_04-b21.

**Comment by Rogach** [comment]
The sys.exit is completely expected in the tests, see [StrangeTest](https://github.com/Rogach/scallop/blob/master/src/test/scala/StrangeTest.scala) (I needed to test that it does exit on `--help` and friends, as well as output).

I use Java 1.6u26, maybe that's the problem. I'll try to look into it now.

**Comment by akisaarinen** [comment]
Yeah, you're right, `sys.exits` are to be expected in that test (and good that you have such a test). 

In my case the whole SBT process exits when running the `test` target. I get returned back to the command-line prompt and have to start SBT again. That shouldn't be happening, right? My hypothesis is that `sys.exit` gets called in some test that should be passing, and the `sys.exit` isn't catched properly (because it wasn't expected), and hence the whole SBT quits.

**Comment by Rogach** [comment]
My guess would be that for some reason it fails to switch to proper SecurityManager (which should be trapping the exit). But I can't verify that yet.

Seems unlikely that other tests fail, since only two tests use those quirky features.

**Comment by Rogach** [comment]
Seems that it is really some transitive problem with sbt - I have automated nuke/clone/test/repeat cycle running, one of 5 runs failed on those 3 tests.

**Comment by Rogach** [comment]
Cycle finished, 3 out of 20 runs failed. That's veeery strange.

**Comment by Rogach** [comment]
In my case, only OptionNameGuessing and Usability tests are failing (only tests that use name guessing). I'll try to roll back a bit, and test previous versions - maybe there is some regression.

**Comment by Rogach** [comment]
Ok, but those tests problems seem to be orthogonal problem to me (I opened a separate issue, #23). Do error messages work properly for you now?

**Comment by akisaarinen** [comment]
Sorry for taking a while to respond. I built the newest master and yes, the error messages for incorrectly formatted fields seem to work correctly now. 

The error message for missing trailing argument could be better, though:

```
Error: org.rogach.scallop.exceptions.OptionParseException: Failed to parse the trailing argument list: ''
```

Maybe throwing a specific exception could help, so one could display an application-specific error message about the missing trailing argument (in my case that would be something like "Output file is missing").

It worries me quite a bit, though, that the code is randomly crashing in your tests.

**Comment by Rogach** [comment]
About OptionParseException - seems a name is a bit ambiguous. As I said, I can't return meaningful information, since parsing of trailing arguments uses backtracking, and if it fails, I have no idea why.

In your specific case, though, you can try this solution:

``` scala
val conf = new LazyScallopConf(args) {
  // option definitions
}
conf.initialize {
  case OptionParseException(args) => if (args.isEmpty) ... // no args were supplied.
}
```

Yes, those tests worry me a lot as well.

**Comment by Rogach** [comment]
Well, I published `0.6.1` version - for what it's worth. I'll think a bit more about getting meaningful errors for trailing args, but progress there seems hard. I'm closing this issue for now.
